cachi2/core/package_managers/yarn_classic/project.py

```diff
@@ -0,0 +1,294 @@
+"""
+Parse the relevant files of a yarn project.
+
+It also provides basic utility functions. The main logic to resolve and prefetch
+the dependencies should be implemented in other modules.
+"""
+
+import json
+import logging
+from dataclasses import dataclass
+from typing import Any, Literal, NamedTuple, Optional, Union
+
+import yaml
+from pyarn import lockfile
+
+from cachi2.core.errors import PackageRejected
+from cachi2.core.rooted_path import RootedPath
+
+log = logging.getLogger(name=__name__)
+
+
+DEFAULT_CACHE_FOLDER = "./.yarn/cache"
+
+
+@dataclass
+class _CommonConfigFile:
+    """A base class for representing a config file.
+
+    :param path: the path to the config file, relative to the request source dir
+    :param data: the raw data for the config file content
+    """
+
+    _path: RootedPath
+    _data: dict[str, Any]
+
+    @property
+    def data(self) -> dict[str, Any]:
+        return self._data
+
+    @property
+    def path(self) -> RootedPath:
+        return self._path
+
+    @property
+    def config_type(self) -> Literal["package_json", "yarnrc", "yarnlock"]:
+        if isinstance(self, PackageJson):
+            return "package_json"
+        elif isinstance(self, YarnRc):
+            return "yarnrc"
+        elif isinstance(self, YarnLock):
+            return "yarnlock"
+        else:
+            raise  # TODO: something
+
+    @classmethod
+    def from_file(cls, path: RootedPath) -> "_CommonConfigFile":
+        """Parse the content of a config file."""
+        return cls(path, {})
+
+    def write(self) -> None:
+        """Write the data to the config file."""
+        with self.path.path.open("w") as f:
+            repr(self.data)
+            f.write("\n")
+
+
+class PackageJson(_CommonConfigFile):
+    """A package.json file.
+
+    This class abstracts the underlying attributes and only exposes what
+    is relevant for the request processing.
+    """
+
+    @property
+    def package_manager(self) -> Optional[str]:
+        """Get the package manager string."""
+        return self.data.get("packageManager")
+
+    @package_manager.setter
+    def package_manager(self, package_manager: str) -> None:
+        """Set the package manager string."""
+        self.data["packageManager"] = package_manager
+
+    @property
+    def install_config(self) -> Optional[dict[str, Any]]:
+        """Get the installConfig dict."""
+        return self.data.get("installConfig", "")
+
+    @classmethod
+    def from_file(cls, path: RootedPath) -> "PackageJson":
+        """Parse the content of a package.json file."""
+        try:
+            with path.path.open("r") as f:
+                package_json_data = json.load(f)
```

Comment:
```suggestion
package_json_data = json.loads(path.path.read_text())
```

Reply:
Nice! I need to get more familiar with `Path`...

Reply:
Note that this is only usable in cases where we **have to** load the whole file from disk to memory, in all other cases a context manager is preferred, that's why I prefer context managers for file ops, because the reads always the same, no surprises :) .

Reply:
Updated